Finite Volume PorousFlow

cpgr opened the pull request:
This is the initial work to support FV in the PorousFlow module.

@loganharbour, would you mind taking a look at this when you get chance before I go too far down the rabbit hole. 😆 I think this was the minimum I could do to get a working FV test problem while having all of the current tests in PorousFlow still working.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1061084/) on 0e2ea04 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21172/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21172/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ed91e1d359fd3b10c258835ee1d38105a7fc8140`

cpgr commented:
Hmm, gcc doesn't like this. It looks like I need lots of `this->template`'s in the code.

cpgr commented:
> Hmm, gcc doesn't like this. It looks like I need lots of `this->template`'s in the code.

Luckily not too many `this->templates` were required!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1061238/) on a652b78 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21172/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1061247/) on a652b78 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ed91e1</th>
      <th colspan=4>#21172 a652b7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.59%</td>
      <td>82.58%</td>
      <td>-0.00%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75517</td>
      <td>75517</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15922</td>
      <td>15926</td>
      <td>+4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21172/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21172/coverage/framework)


## Modules coverage

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>ed91e1</th>
      <th colspan=4>#21172 a652b7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.28%</td>
      <td>96.39%</td>
      <td>+0.11%</td>
      <td>98.24%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10532</td>
      <td>10616</td>
      <td>+84</td>
      <td>503</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>407</td>
      <td>398</td>
      <td>-9</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21172/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21172/coverage/porous_flow)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21172/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

mangerij commented:
Hi, I noticed Ferret failed here but when I looked at CIVET, I saw that all tests passed. Anyone know why?

loganharbour commented:
> Hi, I noticed Ferret failed here but when I looked at CIVET, I saw that all tests passed. Anyone know why?

Unrelated issue. Should be good now.

> @loganharbour, would you mind taking a look at this when you get chance before I go too far down the rabbit hole. 

👍🏼 on my list.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit d4dc9d64d1a4954446c74e2f92e7f338593fcecb
+Subproject commit cae3f7be33c009673de3e429855e3d2432398962
```
**Caution!** This contains a submodule update

cpgr replied:
Not sure what happened here...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1075339/) on eeb4cdd wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

cpgr committed:
Generalise PorousFlowDictator to allow FV variables

Refs #21275

cpgr committed:
Add coupledGenericDofValue

Refs #21275

cpgr committed:
Template materials that provide pressure and saturation

Refs #21275

cpgr committed:
Template temperature material

Refs #21275

cpgr committed:
Template mass fraction materials

Refs #21275

cpgr committed:
Template fluid property materials

Refs #21275

cpgr committed:
Template porosity materials

Refs #21275

cpgr committed:
Template permeability materials

Refs #21275

cpgr committed:
Template relative permeability materials

Refs #21275

cpgr committed:
Generalise PorousFlowJoiner to deal with AD materials

Refs #21275

cpgr committed:
Add AD saturation method to support AD PorousFlow materials

Refs #21275

cpgr committed:
Update FluidState material derivatives to deal with AD variables

Refs #21275

cpgr committed:
Add first FVKernels to PorousFlow

Refs #21275

cpgr committed:
Pressure pulse test of FV PorousFlow objects

The first test to demonstrate that the FV kernels and templated
materials are working correctly.

Refs #21275

cpgr committed:
Template PorousFlowPropertyAux

Refs #21275

cpgr committed:
Make hysteretic rel perms non-AD for now

Eventually these can be made to work with AD variables, but for
now thos allows everything to compile and all tests to pass.

Refs #21275

cpgr committed:
Documenting the use of templating for AD/non-AD

Brief documentation describing how to make classes templated on
the bool is_ad to support both AD and non-AD variables.

Refs #21275

cpgr committed:
Documentation for PorousFlow FVkernels

Refs #21275

cpgr commented:
I've just realised that it doesn't seem like there is any way to find the documentation (https://mooseframework.inl.gov/docs/PRs/21172/site/automatic_differentiation/templated_objects.html) I wrote. It follows on fromhttps://mooseframework.inl.gov/docs/PRs/21172/site/automatic_differentiation/ but I can't find out where this links from. It should possibly be listed in 'Application development'?

GiudGiud commented:
@cpgr the AD page is linked to from some of the systems using AD
https://mooseframework.inl.gov/syntax/Kernels/ or https://mooseframework.inl.gov/syntax/Materials/ for example

cpgr commented:
> @cpgr the AD page is linked to from some of the systems using AD
> https://mooseframework.inl.gov/syntax/Kernels/ or https://mooseframework.inl.gov/syntax/Materials/ for example

Thanks! I was wondering how anyone would find this.

cpgr commented:
> @cpgr I was wondering if you looked at our porous Navier-Stokes slides at https://mooseframework.inl.gov/slides/index.html#/ whether you could tell me what the difference between that and the equations solved by the `porous_flow` module is? I'm hoping there is a difference haha, and if so we should think about documenting that somewhere so our users know where to go for what application/modeling-assumptions is relevant to them.

The governing equations for the `porous_flow` module are outlined here -https://mooseframework.inl.gov/modules/porous_flow/governing_equations.html

They are basically conservation of mass of a fluid component + Darcy's law.

I think the main difference would be that this module is set up for more classical reservoir engineering cases, and can model an arbitrary number of fluid phases (most common use case would be 2 I imagine - a gas phase and a liquid phase), but each phase can have an arbitrary number of fluid components (eg, the liquid can be water + salt + dissolved gas component etc). 

The FV stuff here (which is basically an FV equivalent to the FE objects already in this module) is to make it much easier to model cases where there are discontinuities in one of the variables, such as gas saturation , due to the reservoir properties.

lindsayad commented:
It's fun learning about this. A really dumb question that I haven't answered for myself yet: if you were to model a single phase single component problem, what would your nonlinear variable(s) be? Is it just the pressure?

cpgr commented:
> 

Yes

cpgr commented:
I've been using this for a little while now and it all seems to be working.

The only part I'm not certain on is in `Coupleable.C`. I needed to template `coupledDofValues` to template all the porous flow materials, so added
```
template <>
const GenericVariableValue<false> &
Coupleable::coupledGenericDofValue<false>(const std::string & var_name, unsigned int comp) const
{
  return coupledDofValues(var_name, comp);
}

template <>
const GenericVariableValue<true> &
Coupleable::coupledGenericDofValue<true>(const std::string & var_name, unsigned int comp) const
{
  return adCoupledValue(var_name, comp);
}
```

but I'm not sure on the AD one.

lindsayad commented:
Yea that's probably not quite right. The dof values should return all the degree of freedom values in the solution vector associated with a given element, whereas the coupled value api's return the finite element solution evaluated at the element quadrature points

cpgr commented:
I couldn't find an `adCoupledDofValues` method - before I try and make one can I check that I'm not missing anything?

lindsayad commented:
you're not missing anything 😄

cpgr committed:
Add adCoupledDofValues method

Also implement adDofValuesNeighbor in variable class for
consistency with coupledDofValues.

Refs #21275

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -1739,6 +1739,23 @@ Coupleable::coupledArrayDofValues(const std::string & var_name, unsigned int com
+  if (!_c_is_implicit)
+    mooseError("Not implemented");
+
+  if (!_coupleable_neighbor)
```
I think you’ve got these swapped

lindsayad replied:
In the two lines below

cpgr replied:
Hehehe - I originally had `if(_coupleabe_neighbor)` but swapped it to `if(!_coupleabe_neighbor)` to match the existing methods and forgot to swap which one got returned.

GiudGiud commented:
@cpgr are you waiting for a review on this?

cpgr commented:
> @cpgr are you waiting for a review on this?

Yeah, but no rush. This was to check that what I have done so far is ok. I know that there is a lot of work to make it work with all of the exiting PorousFlow objects, I just don't have time until next month to keep going. 

If you or @loganharbour could just take a quick look and let me know if there is anything that I am not doing right that would be great.

GiudGiud commented:
Sounds good. I ll add it to my list and I think @lindsayad will want to have a look as well since this is FV

lindsayad commented:
Ah I had no idea you were waiting on a review. Sorry @cpgr !

lindsayad commented:
I should have taken the fact that you marked it as ready for review as a cue haha

cpgr commented:
No worries guys!

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -521,6 +521,20 @@ Coupleable::vectorTagValueHelper(const std::string & var_name,
+const GenericVariableValue<true> &
+Coupleable::coupledGenericDofValue<true>(const std::string & var_name, unsigned int comp) const
+{
+  return adCoupledValue(var_name, comp);
```
This doesn't look right, or at least I don't think this would be what the user would expect. I think what you would like to be returning from this is an array where the value part of the dual number is the dof value and the derivative part of the dof value has a single value of one at the corresponding dof index. Is that right? If so, maybe I could work to help you implement that

lindsayad replied:
Oh wait... you already have this... so I think this just needs to be
```c++
  return adCoupledDofValues(var_name, comp);
```
Not sure why I'm unable to make a suggestion here :frowning:

cpgr replied:
Oops!

lindsayad commented on modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C:
```diff
@@ -155,48 +172,85 @@ PorousFlowSingleComponentFluid::initQpStatefulProperties()
+    else
+    {
+      // Density and viscosity, and derivatives wrt pressure and temperature
+      Real rho, drho_dp, drho_dT, mu, dmu_dp, dmu_dT;
+      _fp.rho_mu_from_p_T(MetaPhysicL::raw_value(_porepressure[_qp][_phase_num]) *
+                              _pressure_to_Pascals,
+                          MetaPhysicL::raw_value(_temperature[_qp]) + _t_c2k,
+                          rho,
+                          drho_dp,
+                          drho_dT,
+                          mu,
+                          dmu_dp,
+                          dmu_dT);
+      (*_density)[_qp] = rho;
+      (*_ddensity_dp)[_qp] = drho_dp * _pressure_to_Pascals;
+      (*_ddensity_dT)[_qp] = drho_dT;
+      (*_viscosity)[_qp] = mu / _pressure_to_Pascals / _time_to_seconds;
+      (*_dviscosity_dp)[_qp] = dmu_dp / _time_to_seconds;
+      (*_dviscosity_dT)[_qp] = dmu_dT / _pressure_to_Pascals / _time_to_seconds;
```
A general comment I have is that these dy_dx quantities will themselves be functions of the nonlinear degrees of freedom. I have been slowly trying to add more emphasis to the the 5-parameter AD overloads. Just something to maybe think about

cpgr replied:
I'll have to take a look at these - I remember getting a GitHub notification about them but didn't really take a good look

lindsayad reviewed (changes requested):
Overall this looks really good, and I'm sorry I've taken so long to review. We need to get you and Andy added to the CCB. We've been making a lot of changes to the FV system recently and I hope we still support the work you've done here. We'll see when we re-trigger testing

lindsayad commented on modules/porous_flow/test/tests/pressure_pulse/tests:
```diff
@@ -195,5 +195,15 @@
+    issues = '#21275'
+    design = 'porous_flow/tests/pressure_pulse/pressure_pulse_tests.md'
+    requirement = 'PorousFlow shall correctly simulate the transient evolution of a pressure pulse in 1D using 1 phase physics with FV variables'
+  []
```
Might be good to see a Jacobian test as well

cpgr committed:
adCoupledValue -> adCoupledDofValues

Refs #21172

cpgr commented:
Something has obviously changed - the flux kernel I was using no longer works! Any hints you could give me?

lindsayad commented:
Hmm I think with @grmnptr's recent changes to remove ghost cells this might have plunged your neighbor material property evaluations when you're on a boundary into oblivion. What I would suggest, if you're wiling to wait, is to let #22130 go in, and then I can come push a commit to your PR updating your flux kernel. Would that be ok? I know you've already waited a long time, so if you're eager to have this in, then I can probably come up with a patch now

cpgr commented:
Waiting is fine!

lindsayad commented:
Alrighty, I'll re-save this and come back once #22130 is merged

lindsayad commented:
Ok #22130 is in. Want the follow-on #23233 to get through and with those two I have a pretty good idea about where I want to take this. I don't think it'll be a massive re-write like I initially thought

cpgr commented:
Let me know if there is anything I can do!

lindsayad commented:
You've done plenty @cpgr. It's up to us to make sure your hard work doesn't go to waste

lindsayad committed:
Make FV porous flow work after removal of ghost cells

Now that we've removed ghost cells, we can no longer index into
them on boundaries

lindsayad commented:
Ok @cpgr I finally got around to it! Sorry it took forever! I rebased onto a more recent MOOSE commit and then added one of my own. The important change is that we've removed ghost cells; so you have to be careful to avoid indexing into data that logically corresponds to a ghost cell, e.g. a "neighbor" when you are on a boundary face. In order to circumvent this change, I added a new `FVFluxBC` class that mimics the `FVFluxKernel` class you created but makes sure to avoid that out-of-bounds indexing. 

I tried to read about the porous flow module a bit, but I think it would be best if you looked at that new BC class and decided whether there might be a way to generalize it to more phases like you have for the flux kernel. Anyway hopefully this is helpful and you can develop on top of this and we can get this merged soon!

cpgr commented:
Thanks so much @lindsayad. I'll take a look over the next few days

cpgr commented:
I'm trying to understand why making a new flux bc fixes this? Does a FVFluxKernel not evaluate on the boundary faces anymore, meaning that a flux in/out must be added using a BC?

lindsayad commented:
If a FVFluxBC object exists on a boundary, then the FVFluxKernel does not execute on that boundary. Here is the logic
```c++
bool
FVFluxKernel::skipForBoundary(const FaceInfo & fi) const
{
  // Boundaries to avoid come first, since they are always obeyed
  if (avoidBoundary(fi))
    return true;

  // We're not on a boundary, so in practice we're not 'skipping'
  if (!onBoundary(fi))
    return false;

  // Blanket forcing on boundary
  if (_force_boundary_execution)
    return false;

  // Selected boundaries to force
  for (const auto bnd_to_force : _boundaries_to_force)
    if (fi.boundaryIDs().count(bnd_to_force))
      return false;

  // If we have flux bcs then we do skip
  const auto & flux_pr = _var.getFluxBCs(fi);
  if (flux_pr.first)
    return true;

  // If we don't have flux bcs *and* we do have dirichlet bcs then we don't skip. If we don't have
  // either then we assume natural boundary condition and we should skip
  return !_var.getDirichletBC(fi).first;
}
```

lindsayad commented:
One would hope. The reason I added it is for cell gradients. Cell gradients use Green-Gauss for their calculation so they need face values. If a Dirichlet BC exists then it will use that as the face value; if not, then by default we will use a two term extrapolation to determine the boundary face value based on the cell gradient (so this is actually an implicit calculation) and the cell center value. So practically speaking this extrapolated value probably won't match the Dirichlet value exactly, but hopefully it's close and is only an error of O(h^2).

You are certainly welcome to remove that code if you are happy with the results as they are

cpgr commented:
> One would hope. The reason I added it is for cell gradients. Cell gradients use Green-Gauss for their calculation so they need face values. If a Dirichlet BC exists then it will use that as the face value; if not, then by default we will use a two term extrapolation to determine the boundary face value based on the cell gradient (so this is actually an implicit calculation) and the cell center value. So practically speaking this extrapolated value probably won't match the Dirichlet value exactly, but hopefully it's close and is only an error of O(h^2).
> 
> You are certainly welcome to remove that code if you are happy with the results as they are

cpgr commented:
oops - accidentally closed this!

cpgr committed:
Generalise FV advective flux bc

Refs #21275

cpgr commented:
The missing markdown documentation I can fix. The failing `Modules parallel` test throws an error when it runs the `--n-threads=2` tests

```
A 'FVBoundaryCondition' object already exists with the name 'left_diri'.
```

Does this just need an
```
if (_tid == 0)
 add Dirichlet BC here
```

lindsayad commented:
Yes; that was a stupid oversight by me.

lindsayad reviewed (changes requested):
I'd still like to see a Jacobian test, but otherwise this is good to go as far as I'm concerned.

cpgr committed:
Jacobian test of FVAdvectiveFlux and FVMassTimeDerivative

Refs #21275

cpgr commented:
> I'd still like to see a Jacobian test, but otherwise this is good to go as far as I'm concerned.

I just added a Jacobian test of the FV kernels

lindsayad commented on modules/porous_flow/test/tests/jacobian/tests:
```diff
@@ -1215,4 +1215,15 @@
     design = 'porous_flow/porosity.md PorousFlowPorosityLinear.md'
   []
+
+  [./fv_flux]
```
```suggestion
  [fv_flux]
```

lindsayad commented on modules/porous_flow/test/tests/jacobian/tests:
```diff
@@ -1215,4 +1215,15 @@
+    ratio_tol = 1e-7
+    cli_args = '-mat_fd_type ds Executioner/num_steps=1'
+    threading = '!pthreads'
+    requirement = 'PorousFlow shall compute all Jacobian entries of FV Kernels for flux and mass conservation'
```
```suggestion
    requirement = 'PorousFlow shall compute all Jacobian entries of FV Kernels for flux and mass conservation.'
```
The SQA tsars have told me that requirements should be complete sentences

lindsayad commented on modules/porous_flow/test/tests/jacobian/tests:
```diff
@@ -1215,4 +1215,15 @@
+    requirement = 'PorousFlow shall compute all Jacobian entries of FV Kernels for flux and mass conservation'
+    issues = '#21275'
+    design = 'FVPorousFlowMassTimeDerivative.md FVPorousFlowAdvectiveFlux.md'
+  [../]
```
```suggestion
  []
```

lindsayad reviewed (approved):
Above and beyond on that last commit @cpgr :smile:

cpgr commented:
> Above and beyond on that last commit @cpgr 😄

`sed` and some search and replace!

GiudGiud commented:
@lindsayad do you need an external review on your commit or are we below the "significance" threshold?

lindsayad commented:
I don't think it's very significant, so I don't think it needs it. Feel free to review if you wish obviously, but I think we're good to move forward without more review

GiudGiud commented on modules/porous_flow/src/fvbcs/FVPorousFlow1PhaseAdvectiveFluxBC.C:
```diff
@@ -0,0 +1,83 @@
+{
+  InputParameters params = FVFluxBC::validParams();
+  RealVectorValue g(0, 0, -9.81);
+  params.addParam<RealVectorValue>("gravity", g, "Gravity vector. Defaults to (0, 0, -9.81)");
```
how often do you do 2D @cpgr ?

this default isnt great for 2D, I d almost want to error if this is used in 2D

cpgr replied:
Yeah, I just used the same default we have in the other porous flow objects that exist. 

I guess we could either not have a default, or do something fancier like check the mesh dimension and then set an appropriate value?

GiudGiud replied:
both are an option. It s not tied to this object then, can be figured out later

GiudGiud commented:
@cpgr please let me know when you want this merged

lindsayad commented:
I bet he's ready to have it merged now 😄

cpgr commented:
Haha - what's a few extra days!

cpgr commented:
Thanks @lindsayad for your help getting this through!

lindsayad commented:
You're welcome! Thanks for your patience in the process!
